File: Polinomio Taylor.py
def factorial(n):
    """
    Calcula el factorial de un número entero de forma iterativa.
    Args:
        n (int): Número del cual se quiere calcular el factorial.
    Returns:
        int: El factorial de n.
    """
    resultado = 1
    for i in range(2, n + 1):
        resultado *= i
    return resultado


def polinomio_taylor_seno(x, n):
    """
    Calcula la aproximación del seno usando el polinomio de Taylor centrado en x=0.
    
    Args:
        x (float): Valor en el cual se evalúa el seno.
        n (int): Número de términos del polinomio de Taylor a usar.
    
    Returns:
        float: Aproximación del seno de x.
    """
    seno_aprox = 0.0
    potencia = x  # x^(2k+1), empieza en x^1
    fact = 1      # (2k+1)!, empieza en 1!
    signo = 1     # (-1)^k, empieza positivo

    for k in range(n):
        termino = signo * potencia / fact
        seno_aprox += termino

        # Preparar el siguiente término
        signo *= -1
        potencia *= x * x  # x^(2k+3) = x^(2k+1) * x^2
        fact *= (2 * k + 2) * (2 * k + 3)  # (2k+3)! = (2k+1)! * (2k+2)*(2k+3)

    return seno_aprox
print(polinomio_taylor_seno(0, 5))        # 0.0
print(polinomio_taylor_seno(1, 10))       # ≈ 0.8414709848078965
print(polinomio_taylor_seno(-1, 7))       # ≈ -0.8414709848078965
print(polinomio_taylor_seno(3.14159, 15)) # ≈ 0.0000026535897933527

File: polinomio taylor.kt
fun factorial(n: Int): Long {
    var resultado = 1L
    for (i in 2..n) {
        resultado *= i
    }
    return resultado
}

fun polinomioTaylorSeno(x: Double, n: Int): Double {
    var senoAprox = 0.0
    var potencia = x          // x^(2k+1), comienza en x^1
    var factorial = 1.0       // (2k+1)!, comienza en 1!
    var signo = 1.0           // (-1)^k

    for (k in 0 until n) {
        val termino = signo * potencia / factorial
        senoAprox += termino

        // Preparar siguiente término
        signo *= -1
        potencia *= x * x
        factorial *= (2 * k + 2) * (2 * k + 3)
    }

    return senoAprox
}
fun main() {
    println(polinomioTaylorSeno(0.0, 5))         // 0.0
    println(polinomioTaylorSeno(1.0, 10))        // ≈ 0.8414709848078965
    println(polinomioTaylorSeno(-1.0, 7))        // ≈ -0.8414709848078965
    println(polinomioTaylorSeno(3.14159, 15))    // ≈ 0.0000026535...
}
